cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/TransactionHttpHandler.java

```diff
@@ -319,45 +303,49 @@ private boolean initializePruningDebug(HttpResponder responder) {
     }
 
     synchronized (this) {
-      if (debugClazz == null || debugObject == null) {
-        try {
-          this.debugClazz = getClass().getClassLoader()
-            .loadClass("org.apache.tephra.hbase.txprune.InvalidListPruningDebug");
-          this.debugObject = debugClazz.newInstance();
-          for (Map.Entry<String, String> entry : cConf) {
-            hConf.set(entry.getKey(), entry.getValue());
-          }
-          Method initMethod = debugClazz.getMethod("initialize", Configuration.class);
-          initMethod.setAccessible(true);
-          initMethod.invoke(debugObject, hConf);
-        } catch (ClassNotFoundException | IllegalAccessException | InstantiationException ex) {
-          LOG.debug("InvalidListPruningDebug class not found. Pruning Debug endpoints will not work.", ex);
-          this.debugClazz = null;
-        } catch (NoSuchMethodException | InvocationTargetException ex) {
-          LOG.debug("Initialize method was not found in InvalidListPruningDebug. Debug endpoints will not work.", ex);
-          this.debugClazz = null;
-        }
+      if (pruningDebug != null) {
+        return true;
+      }
+
+      // Copy over both cConf and hConf into the pruning configuration
+      Configuration configuration = new Configuration();
+      configuration.clear();
+      // First copy hConf and then cConf so that we retain the values from cConf for any parameters defined in both
+      copyConf(configuration, hConf);
+      copyConf(configuration, cConf);
+
+      try {
+        @SuppressWarnings("unchecked")
+        Class<? extends InvalidListPruningDebug> clazz =
+          (Class<? extends InvalidListPruningDebug>) getClass().getClassLoader().loadClass(PRUNING_TOOL_CLASS_NAME);
+        this.pruningDebug = clazz.newInstance();
+        pruningDebug.initialize(configuration);
+      } catch (ClassNotFoundException | InstantiationException | IllegalAccessException |
+        ClassCastException | IOException e) {
+        LOG.debug("Not able to instantiate pruning debug class", e);
+        pruningDebug = null;
```

Comment:
if it fails, you must respond with an error and return false. That is what the callers expect.

Reply:
Good catch. I had missed that.